src/udev/udev-node.c

```diff
@@ -188,28 +201,116 @@ static int link_find_prioritized(sd_device *dev, bool add, const char *stackdir,
         return 0;
 }
 
+static int init_link_semaphores(const char *path) {
+        key_t key = ftok(path, 0);
+        int semid;
+
+        semid = semget(key, N_SEMAPHORES, 0600|IPC_CREAT|IPC_EXCL);
+        if (semid != -1) {
+                unsigned short val[N_SEMAPHORES];
+                int i;
+                struct sembuf dummy_op[]  = {
+                        { .sem_op = -1, },
+                        { .sem_op = 1, },
+                };
+
+                for (i = 0; i < N_SEMAPHORES; i++)
+                        val[i] = 1;
+                if (semctl(semid, 0, SETALL, val) == -1)
+                        return log_error_errno(-errno, "Failed to initialize semaphores: %m");
+
+                /* Dummy semop to set sem_otime */
+                if (semop(semid, dummy_op, (sizeof(dummy_op)/sizeof(*dummy_op))) == -1)
+                        return log_error_errno(-errno, "Failed to set sem_otime: %m");
+                return semid;
+        } else {
+                const unsigned int RETRIES = 10, SLEEP_US = 10000;
+                unsigned int i;
+
+                semid = semget(key, N_SEMAPHORES, 0);
+                if (semid == -1)
+                        return log_error_errno(-errno, "Failed to get semaphore set: %m");
+
+                for (i = 0; i < RETRIES; i++) {
+                        struct semid_ds ds;
+
+                        /* Wait for initialization to finish */
+                        if (semctl(semid, 0, IPC_STAT, &ds) != -1 &&
+                            ds.sem_otime != 0)
+                                return semid;
+                        usleep(SLEEP_US);
+                }
+                return log_error_errno(SYNTHETIC_ERRNO(ENOENT),
+                                       "Semaphore set not initialized after %d us",
+                                       RETRIES * SLEEP_US);
+        }
+}
+
+static unsigned short get_sema_index(const char *link) {
+        static const unsigned char seed[16] = { 0x6b, 0xb0, 0xb1, 0x28,
```

Comment:
why "unsignec char"? siphash24_init expects an uint8_t[] array there... which is of course the same, but let's not create needless confusion